### ISSUE TITLE
MultiRaft can deadlock between sendEvent and attempting to write to callbackChan

### ISSUE THREAD

**Issue by tschottdorf** [cockroachdb/cockroach#1185]
From a clean master at the time of writing (@de153f0f619c9b837cbb4481dccbef44da34b80b):

- change the put test so that it leaves the cluster running when it's done (add `<-stopper`)
- TESTS=TestPut TESTTIMEOUT=24h TESTFLAGS="-d 65s" acceptance/run.sh
- wait until it's done
- check the status endpoints for the three nodes. for most runs, I get the following:
  - range descriptors indicate three (may vary slightly) ranges (via ./cockroach range ls on any of the nodes)
  - first and third node show those on the web frontend (well, they show "3" ranges)
  - second node has never split any ranges - it only shows one range (debugging shows that it's the initial range spanning everything)
  - stats indicate furthermore that that node doesn't hold any data (14 keys and handful more of values).

In my tests with a lot of debug statements (in particular for EndTransaction calls) I don't see any such calls logged after the replica changes. The node does however receive Put requests (though presumably not all of them), these fail because store.GetRange doesn't know anything but the old incorrect range 1 (""-"\xff\xff").

Something really weird is going on here; the node seems to be left out.

```
root@e2d00c9b91b3:/# du -sh data*
196M	data0
100K	data1
195M	data2
```

**Comment by tschottdorf** [comment]
helpful snippet:
`for i in 0 1 2; do export CONTAINER=$(docker ps --no-trunc | grep data${i} | awk '{print $1}') && docker logs -f ${CONTAINER} 2>&1 | tee tmp$i.log & done`

**Comment by tamird** [comment]
This is even better, when running with 5 nodes, nodes 0-2 are fine, but 3 and 4 are completely idle.

**Comment by tschottdorf** [comment]
I can confirm that the second node stops processing Raft commands right after the EndTransaction call with the SplitTrigger which adds the third node's replica. Processing does not block inside of `rng.processRaftCommand` (checked via logging) so that I'll look next whether MultiRaft is receiving messages.

```
19:20 $ grep -c processRaftCommand tmp*
tmp0.log:46998
tmp1.log:16
tmp2.log:45993
```

the last command which the range executes is the ChangeReplica EndTransaction.

>
EndTransaction header:<timestamp:<wall_time:1432855147514056465 logical:0 > cmd_id:<wall_time:1432855147514049234 random:6640337109912895430 > key:"" end_key:"" user:"root" replica:<node_id:"\001" store_id:"\001" attrs:<> > raft_id:1 txn:<name:"change replicas of 1" key:"" id:"\233\2169\224\234\005O_\207\205B\343\200\213\241n" priority:759455126 isolation:SERIALIZABLE status:PENDING epoch:0 timestamp:<wall_time:1432855147514056465 logical:0 > orig_timestamp:<wall_time:1432855147514056465 logical:0 > max_timestamp:<wall_time:1432855147764056465 logical:0 > certain_nodes:<> > read_consistency:CONSISTENT > commit:true internal_commit_trigger:<change_replicas_trigger:<node_id:"\003" store_id:"\003" change_type:ADD_REPLICA updated_replicas:<node_id:"\001" store_id:"\001" attrs:<> > updated_replicas:<node_id:"\002" store_id:"\002" attrs:<attrs:"ssd" > > updated_replicas:<node_id:"\003" store_id:"\003" attrs:<attrs:"ssd" > > > intents:"\000\000\000k\000\001rdsc" >

**Comment by tschottdorf** [comment]
I can't reproduce this on my Linux machine (5 runs, all with the second node behaving like it ought with the exception of #1189). Maybe a timing issue - the Docker stuff seems to be running swifter, though the Put throughput is about the same.

**Comment by tschottdorf** [comment]
Ok, I think I'm going to be able to get to the bottom of this today - MultiRaft is blocked sending the event after the membership committed event, which makes it out of Raft processing but then gets stuck executing the callback (see multiraft.go:879ish).

**Comment by tschottdorf** [comment]
It's stuck passing the func into the `callbackChan`. Presumably the following deadlock happens:

- MultiRaft sends the commit event for the Membership change
- `store.processRaft` takes it and begins processing it
- the next command commits and triggers another sendEvent, but this blocks since `store.processRaft` isn't ready for another select
- the Membership change was applied to the range, and the store now tries to execute the callback
- the callback tries to write to `callbackChan`, but that is consumed by the MultiRaft loop, which is currently waiting for `store.processRaft` to consume from the events channel, which it will only do after the callback has completed.

Voila, a veritable deadlock.

**Comment by tschottdorf** [comment]
This bug is uncovered due to the recent addition of the `EventBufferSize` (see the comment in `multiraft.go:73`. I'm going to look for a clean solution to this deadlock.
ping @bdarnell

**Comment by petermattis** [comment]
#1193 might be the same problem as my debugging is indicating raft processing is locking up.

**Comment by bdarnell** [comment]
This is exactly the sort of bug I was hoping to shake out by making the channel unbuffered. :)  Re-adding the channel buffer works around it for now, but any deadlock that arises with an unbuffered channel can also occur when a buffered channel reaches its limit. I think the fix is going to be to undo #1143 (and make the nil channel dance a little more complicated).

**Comment by bdarnell** [comment]
Nevermind that last comment; #1143 is not at fault here. The issues involved here are subtle; I now think that the use of buffered channels here has been buggy all along. The problem is that for a given Ready batch, we must make any necessary calls to ApplyConfChange before calling MultiNode.Advance. If either the event or callback channel is buffered, these calls could get out of order (but if they're not buffered, we have the deadlock seen here). 

The channel-based design isn't really working here. I think it might be best to replace the events and callback channels with a synchronous delegate interface, although I'm concerned about how much the multiraft select loop will now be blocked (I was expecting that we would buffer the events channel in production)
